feat(services/webdav): support redirection when get 302/307 response during read operation

Yansongsongsong opened the pull request:
related issue: #1355

Xuanwo commented on core/src/services/webdav/backend.rs:
```diff
@@ -471,6 +472,79 @@ impl WebdavBackend {
+    // read will send http request for dav access.
+    // sometimes we need to send request to different endpoint when server side
+    // indicate we need to redirect to another url.
+    #[async_recursion]
```
We don't need to do recursion, we can expand the redirection by hand. And we should add a new config value for the redirection limit (maybe default to 3 times).

Yansongsongsong replied:
Hi i have modified it according to the review.

I also did some changes to pass all CI tests. Please help have one look on this if you got time

Yansongsongsong commented on core/src/services/webdav/fixtures/nginx.conf:
```diff
@@ -8,6 +8,23 @@ events {
+        return 302 http://$server_name:8080$request_uri;
+      }
+      client_max_body_size 1024M;
+      # forward all other requests to port 8081
```
```suggestion
      # forward all other requests to port 8080
```

Xuanwo commented on core/src/services/webdav/backend.rs:
```diff
@@ -299,17 +302,81 @@ impl Accessor for WebdavBackend {
+                }
+                StatusCode::FOUND | StatusCode::TEMPORARY_REDIRECT => {
+                    // if server returns redirect HTTP status, then redirect it
+                    let redirected_url = parse_location(resp.headers())?
```
We will need to handle the `stat` operation too.

dtretyakov commented:
@Yansongsongsong are you still working on addressing review comments or you have no time?

Yansongsongsong commented:
> @Yansongsongsong are you still working on addressing review comments or you have no time?

@dtretyakov hi i am still working on that. i may push my change tmr. sorry for late

Yansongsongsong commented:
hi  @Xuanwo , sorry for late. have refactored according to the review. please have one look at this. thank you

Yansongsongsong commented on .github/workflows/service_test_webdav.yml:
```diff
@@ -112,3 +112,38 @@ jobs:
+        run: |
+          mkdir -p /tmp/static
+          mkdir -p /var/lib/nginx
+          chmod a+rw /tmp/static/ # make nginx worker has permission to operate it
```
refactored. pls have one look

Yansongsongsong commented on core/src/services/webdav/backend.rs:
```diff
@@ -299,17 +302,81 @@ impl Accessor for WebdavBackend {
-            StatusCode::OK | StatusCode::PARTIAL_CONTENT => {
-                let meta = parse_into_metadata(path, resp.headers())?;
-                Ok((RpRead::with_metadata(meta), resp.into_body()))
+        let mut remaining_retry_times = 3;
```
refactored. use `retry` to indicated how many times have been tried

Xuanwo commented:
> hi @Xuanwo , sorry for late. have refactored according to the review. please have one look at this. thank you

Sorry for miss this comment, I will take a look today.

Yansongsongsong commented:
> Please resolve the conflict so that we can merge this PR in!

@Xuanwo conflicts resolved. thank you for mentoring.

Xuanwo commented:
Hi, @Yansongsongsong please update branch and try again. Sorry for our fault.
